api/capi/include/nntrainer.h

```diff
@@ -65,9 +72,31 @@ typedef enum {
   ML_LAYER_TYPE_UNKNOWN    /**< Unknown Lyaer */
 } ml_layer_type_e;
 
+/**
+ * @brief Dataset generator callback function for train/valid/test data.
+ * @details The Containers passed will already be allocated with sufficient
+ * space to contain the data by the caller. This function should return a batch
+ * of data in the passed containers.
+ * @note This function can be called multiple times in parallel.
+ * @param[out] input Container to provide input data to the model.
+ * @param[out] label Container to provide label data to the model to train/test.
+ * @param[out] last Container to provide is the data is finished.
+ * @return @c 0 on success. Otherwise a negative error value.
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_INVALID_PARAMETER Invalid Parameter.
+ */
+typedef int (*dataset_generator) (float *input, float *label, bool *last);
```

Comment:
I think there should be a doc about format of the data.

sth like:
1. It should match the shape of specified input
2. data is read row-wise

Reply:
1. I am thinking of adding a `size_t bytes` in this function for this, as this function should know about the size of the allocated memory. Yes, added a comment for shape.
2. Added.

Reply:
Updated `float *input`  and `float *label` to `float **input` and `float **label` to support multiple inputs and outputs.